appsync/aws-appsync-events/src/main/java/com/amplifyframework/aws/appsync/events/EventsChannel.kt

```diff
@@ -98,12 +98,41 @@ class EventsChannel internal constructor(
      * @param authorizer for the publish call. If not provided, the EventChannel publish authorizer will be used.
      * @return result of publish.
      */
-    @Throws(EventsException::class)
+    @Throws(Exception::class)
     suspend fun publish(
         events: List<JsonElement>,
         authorizer: AppSyncAuthorizer = this.authorizers.publishAuthorizer
-    ): PublishResult {
-        TODO("Need to implement")
+    ): PublishResult = coroutineScope {
+        val publishId = UUID.randomUUID().toString()
+        val publishMessage = WebSocketMessage.Send.Publish(
+            id = publishId,
+            channel = name,
+            events = JsonArray(events.map { JsonPrimitive(it.toString()) }),
+        )
+
+        val webSocket = eventsWebSocketProvider.getConnectedWebSocket()
+        val deferredResponse = async { getPublishResponse(webSocket, publishId) }
+
+        val queued = webSocket.sendWithAuthorizer(publishMessage, authorizer)
+        if (!queued) {
+            webSocket.disconnectReason?.toCloseException() ?: ConnectionClosedException()
+        }
+
+        val response = deferredResponse.await()
+        return@coroutineScope when {
+            response is WebSocketMessage.Received.PublishSuccess -> {
+                PublishResult(response.successfulEvents, response.failedEvents)
+            }
+            response is WebSocketMessage.ErrorContainer && response.id == publishId -> {
```

Comment:
`publishId` should have already been checked in `getPublishResponse`.

Reply:
good callout